src/app/services/shared-activated-route/shared-activated-route.service.ts

```diff
@@ -0,0 +1,131 @@
+import { Injectable } from "@angular/core";
+import {
+  ActivatedRoute,
+  Data,
+  NavigationEnd,
+  Params,
+  Router,
+} from "@angular/router";
+import { PageComponent } from "@helpers/page/pageComponent";
+import { IPageInfo, isIPageInfo } from "@helpers/page/pageInfo";
+import { withUnsubscribe } from "@helpers/unsubscribe/unsubscribe";
+import {
+  BehaviorSubject,
+  filter,
+  map,
+  mergeMap,
+  Observable,
+  takeUntil,
+} from "rxjs";
+
+/**
+ * Components outside of the router-outlet are unable to cleanly access data
+ * from inside the router-outlet. This affects things like the menu which
+ * exists outside of the current route, and thus cannot access route/query
+ * parameters. To bypass this, we pass the activated route to this service,
+ * and then pass it along to the components which rely on this information.
+ *
+ * Some relevant issues documenting this issue:
+ *
+ * - https://github.com/angular/angular/issues/15004
+ * - https://github.com/angular/angular/issues/11023
+ */
+@Injectable({ providedIn: "root" })
+export class SharedActivatedRouteService extends withUnsubscribe() {
+  private _activatedRoute$: BehaviorSubject<ActivatedRoute>;
+  private currentActivatedRoute: ActivatedRoute;
+
+  public constructor(router: Router, route: ActivatedRoute) {
+    super();
+    this.setActivatedRoute(route);
+    this.observeRouterEvents(router, route);
+  }
+
+  /** Triggers whenever activated route for the page changes */
+  public get activatedRoute$(): Observable<ActivatedRoute> {
+    return this._activatedRoute$;
+  }
+
+  public get data(): Observable<Data> {
+    return this._activatedRoute$.pipe(
+      mergeMap((route): Observable<Data> => route.data)
+    );
+  }
+
+  public get pageInfo(): Observable<IPageInfo> {
+    return this.data.pipe(filter((data): boolean => isIPageInfo(data)));
+  }
+
+  public get params(): Observable<Params> {
+    return this._activatedRoute$.pipe(map((route): Params => route.params));
+  }
+
+  public get queryParams(): Observable<Params> {
+    return this._activatedRoute$.pipe(
+      map((route): Params => route.queryParams)
+    );
+  }
+
+  private observeRouterEvents(router: Router, route: ActivatedRoute): void {
+    const findPageComponent = (
+      _route: ActivatedRoute,
+      depth: number
+    ): ActivatedRoute | undefined => {
+      // Prune non primary routes from search
+      if (_route.outlet !== "primary") {
+        return undefined;
+      }
+
+      // If page component found, return activated route
+      if ((_route.component as any)?.prototype instanceof PageComponent) {
+        return _route;
+      }
+
+      // Hit max depth, return undefined
+      if (depth === 5) {
+        return undefined;
+      }
```

Comment:
Why is this needed?

Arbitrary depths in traversals are a huge code smell for me